hazelcast/src/main/java/com/hazelcast/config/YamlConfigBuilder.java

```diff
@@ -0,0 +1,302 @@
+     * @param locator the configured locator to use
+     */
+    public YamlConfigBuilder(YamlConfigLocator locator) {
+        if (locator == null) {
+            locator = new YamlConfigLocator();
+            locator.locateFromSystemProperty();
+        }
+
+        this.in = locator.getIn();
+        this.configurationFile = locator.getConfigurationFile();
+        this.configurationUrl = locator.getConfigurationUrl();
+    }
+
+    @Override
+    public Config build() {
+        return build(new Config());
+    }
+
+    Config build(Config config) {
+        config.setConfigurationFile(configurationFile);
+        config.setConfigurationUrl(configurationUrl);
+        try {
+            parseAndBuildConfig(config);
+        } catch (Exception e) {
+            throw ExceptionUtil.rethrow(e);
+        }
+        return config;
+    }
+
+    private void parseAndBuildConfig(Config config) throws Exception {
+        YamlMapping yamlRootNode;
+        try {
+            yamlRootNode = ((YamlMapping) YamlLoader.load(in));
+        } catch (Exception ex) {
+            throw new InvalidConfigurationException("Invalid YAML configuration", ex);
+        }
+
+        YamlNode imdgRoot = yamlRootNode.childAsMapping(ConfigSections.HAZELCAST.name().toLowerCase());
+        if (imdgRoot == null) {
+            throw new InvalidConfigurationException("No mapping with hazelcast key is found in the provided configuration");
+        }
+
+        Node w3cRootNode = asW3cNode(imdgRoot);
+        replaceVariables(w3cRootNode);
+        importDocuments(imdgRoot);
+
+        new YamlMemberDomConfigProcessor(true, config).buildConfig(w3cRootNode);
+    }
+
+    private void importDocuments(YamlNode imdgRoot) throws Exception {
+        YamlMapping rootAsMapping = asMapping(imdgRoot);
+        YamlSequence importSeq = rootAsMapping.childAsSequence(ConfigSections.IMPORT.name);
+        if (importSeq == null || importSeq.childCount() == 0) {
+            return;
+        }
+
+        for (YamlNode importNode : importSeq.children()) {
+            String resource = asScalar(importNode).nodeValue();
+            URL url = ConfigLoader.locateConfig(resource);
+            if (url == null) {
+                throw new InvalidConfigurationException("Failed to load resource: " + resource);
+            }
+            if (!currentlyImportedFiles.add(url.getPath())) {
+                throw new InvalidConfigurationException("Cyclic loading of resource '" + url.getPath() + "' detected!");
+            }
+
+            YamlNode rootLoaded;
+            try {
+                rootLoaded = YamlLoader.load(url.openStream());
+            } catch (Exception ex) {
+                throw new InvalidConfigurationException("Loading YAML document from resource " + url.getPath() + " failed", ex);
+            }
+            YamlNode imdgRootLoaded = asMapping(rootLoaded).child(ConfigSections.HAZELCAST.name().toLowerCase());
+
+            replaceVariables(asW3cNode(imdgRootLoaded));
```

Comment:
We may produce NPE inside `replaceVariables` if there is no `hazelcast` section in the loaded YAML. Probably it's better just to perform an early exit, we don't need null check in the `merge` then.

Reply:
Good point, done.